File: packages/vscode-extension/src/chat/conversation-manager.js
/**
 * Pure conversation/tab model for the chat panel (Claude-Code conversation-tabs
 * parity). Holds N named conversations, each owning its own agent session
 * handle, resume id, per-turn reducer state, and title; tracks which one is
 * active. No `vscode` and no child-process coupling — the live AgentChatSession
 * is stored opaquely — so this is fully unit-testable. chat-view.js drives it
 * (spawn/route/dispose); chat-html.js renders the tab bar from list().
 *
 * This is slice 1 of the multi-tab feature: the data layer behind the panel.
 * The webview wiring (per-tab child, tab bar UI) builds on top of it.
 */

class ConversationManager {
  /**
   * @param {object} opts
   * @param {function} [opts.createTurnState] factory for a conversation's
   *        per-turn reducer state (chat-events.createTurnState); injected so
   *        this module stays dependency-free and testable.
   */
  constructor(opts = {}) {
    this._createTurnState =
      typeof opts.createTurnState === "function"
        ? opts.createTurnState
        : () => ({});
    this._conversations = new Map(); // id -> conversation record
    this._order = []; // tab order (array of ids, left→right)
    this._activeId = null;
    this._seq = 0; // monotonic — stable ids/titles even after closes
  }

  _nextId() {
    this._seq += 1;
    return `conv-${this._seq}`;
  }

  /**
   * Create a conversation. Becomes active when `activate` (default) or when no
   * conversation is active yet. Returns the new record.
   */
  create({ title, sessionId = null, activate = true } = {}) {
    const id = this._nextId();
    const conv = {
      id,
      title: title || `Chat ${this._seq}`,
      sessionId: sessionId || null,
      session: null, // opaque AgentChatSession handle (set by chat-view)
      turnState: this._createTurnState(),
      unread: false, // a turn finished while this tab was in the background
      mode: "default", // approval mode (default | acceptEdits | bypassPermissions)
      thinking: "off", // extended thinking (off | on | ultra)
    };
    this._conversations.set(id, conv);
    this._order.push(id);
    if (activate || this._activeId == null) this._activeId = id;
    return conv;
  }

  get(id) {
    return this._conversations.get(id) || null;
  }

  has(id) {
    return this._conversations.has(id);
  }

  activeId() {
    return this._activeId;
  }

  active() {
    return this._activeId ? this.get(this._activeId) : null;
  }

  count() {
    return this._order.length;
  }

  /** Tabs for the UI, in left→right order, flagging the active one. */
  list() {
    return this._order.map((id) => {
      const c = this._conversations.get(id);
      return {
        id: c.id,
        title: c.title,
        sessionId: c.sessionId,
        active: c.id === this._activeId,
        hasSession: !!c.session,
        unread: !!c.unread,
        mode: c.mode || "default",
      };
    });
  }

  /** Activate a tab. Clears its unread flag (you're now looking at it).
   * Returns its record, or null if the id is unknown. */
  switchTo(id) {
    if (!this._conversations.has(id)) return null;
    this._activeId = id;
    const c = this.get(id);
    if (c) c.unread = false;
    return c;
  }

  /**
   * Flag a conversation as having a freshly-completed turn the user hasn't seen
   * yet — the tab bar renders a dot for it (background-tab completion signal).
   * No-op for the active tab (you're already watching it). Returns the record,
   * or null if the id is unknown or it was the active tab (left unflagged).
   */
  markUnread(id) {
    if (id === this._activeId) return null;
    const c = this.get(id);
    if (c) c.unread = true;
    return c || null;
  }

  /**
   * Remove a conversation. If it was active, activate the neighbor to its right
   * (or left if it was last; null if none remain) — the natural tab-close UX.
   * Returns `{ closed, conv?, active }` where `conv` is the removed record (so
   * the caller can stop its child) and `active` is the new active record.
   */
  close(id) {
    if (!this._conversations.has(id)) {
      return { closed: false, active: this.active() };
    }
    const idx = this._order.indexOf(id);
    const conv = this._conversations.get(id);
    this._conversations.delete(id);
    this._order.splice(idx, 1);
    if (this._activeId === id) {
      // After splice, _order[idx] is what used to be the right neighbor.
      this._activeId = this._order[idx] || this._order[idx - 1] || null;
    }
    return { closed: true, conv, active: this.active() };
  }

  setSessionId(id, sessionId) {
    const c = this.get(id);
    if (c) c.sessionId = sessionId || null;
    return c;
  }

  /** Set a conversation's approval mode (takes effect on its next child spawn). */
  setMode(id, mode) {
    const c = this.get(id);
    if (c) c.mode = mode || "default";
    return c;
  }

  /** Set a conversation's extended-thinking level (next child spawn). */
  setThinking(id, thinking) {
    const c = this.get(id);
    if (c) c.thinking = thinking || "off";
    return c;
  }

  setSession(id, handle) {
    const c = this.get(id);
    if (c) c.session = handle || null;
    return c;
  }

  setTitle(id, title) {
    const c = this.get(id);
    if (c && title) c.title = title;
    return c;
  }

  /** Fresh per-turn reducer state for a conversation (e.g. after reset/resume). */
  resetTurnState(id) {
    const c = this.get(id);
    if (c) c.turnState = this._createTurnState();
    return c;
  }

  /** Every live session handle (for dispose — caller stops each). */
  allSessions() {
    return this._order
      .map((id) => this._conversations.get(id).session)
      .filter(Boolean);
  }
}

module.exports = { ConversationManager };

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/ConversationManager.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * Pure conversation/tab model for the JetBrains chat tool window (Claude-Code
 * conversation-tabs parity) — a direct port of the VS Code extension's
 * conversation-manager.js. Holds N named conversations, each owning its own
 * agent session handle, resume id, per-turn reducer state, and title; tracks
 * which one is active. No IntelliJ SDK — the live AgentChatSession is held
 * opaquely as Object — so it compiles + tests with plain {@code javac}.
 *
 * The IntelliJ glue (a future tool-window with a real tab strip) drives this:
 * spawn a child per tab, route events to the owning tab, render the bar from
 * {@link #list()}.
 */
public final class ConversationManager {

    /** One conversation/tab. The session handle + turn state are opaque. */
    public static final class Conversation {
        public final String id;
        public String title;
        public String sessionId;   // resume id (nullable)
        public Object session;     // opaque AgentChatSession handle (nullable)
        public Object turnState;
        public String mode = "default"; // approval mode (default|acceptEdits|bypassPermissions)
        public String thinking = "off"; // extended thinking (off|on|ultra)

        Conversation(String id, String title, String sessionId, Object turnState) {
            this.id = id;
            this.title = title;
            this.sessionId = sessionId;
            this.turnState = turnState;
        }
    }

    /** Tab row for the UI, mirroring list() in the JS model. */
    public static final class TabInfo {
        public final String id;
        public final String title;
        public final String sessionId;
        public final boolean active;
        public final boolean hasSession;

        TabInfo(String id, String title, String sessionId, boolean active, boolean hasSession) {
            this.id = id;
            this.title = title;
            this.sessionId = sessionId;
            this.active = active;
            this.hasSession = hasSession;
        }
    }

    /** Result of close(): whether it removed anything, the removed record, the new active. */
    public static final class CloseResult {
        public final boolean closed;
        public final Conversation conv;    // removed record (nullable)
        public final Conversation active;  // new active record (nullable)

        CloseResult(boolean closed, Conversation conv, Conversation active) {
            this.closed = closed;
            this.conv = conv;
            this.active = active;
        }
    }

    private final Supplier<Object> createTurnState;
    private final Map<String, Conversation> conversations = new HashMap<String, Conversation>();
    private final List<String> order = new ArrayList<String>(); // tab order, left→right
    private String activeId = null;
    private int seq = 0; // monotonic — stable ids/titles even after closes

    public ConversationManager() {
        this(null);
    }

    /** @param createTurnState factory for a conversation's per-turn reducer state (nullable). */
    public ConversationManager(Supplier<Object> createTurnState) {
        this.createTurnState = createTurnState != null ? createTurnState : new Supplier<Object>() {
            public Object get() { return new HashMap<String, Object>(); }
        };
    }

    private String nextId() {
        seq += 1;
        return "conv-" + seq;
    }

    /** Create a conversation; becomes active when {@code activate} or none is active yet. */
    public Conversation create(String title, String sessionId, boolean activate) {
        String id = nextId();
        String t = (title != null && !title.trim().isEmpty()) ? title : ("Chat " + seq);
        Conversation conv = new Conversation(id, t, emptyToNull(sessionId), createTurnState.get());
        conversations.put(id, conv);
        order.add(id);
        if (activate || activeId == null) activeId = id;
        return conv;
    }

    /** Convenience: a fresh active conversation with default title and no resume id. */
    public Conversation create() {
        return create(null, null, true);
    }

    public Conversation get(String id) {
        return conversations.get(id);
    }

    public boolean has(String id) {
        return conversations.containsKey(id);
    }

    public String activeId() {
        return activeId;
    }

    public Conversation active() {
        return activeId != null ? conversations.get(activeId) : null;
    }

    public int count() {
        return order.size();
    }

    /** Tabs in left→right order, flagging the active one. */
    public List<TabInfo> list() {
        List<TabInfo> out = new ArrayList<TabInfo>();
        for (String id : order) {
            Conversation c = conversations.get(id);
            out.add(new TabInfo(c.id, c.title, c.sessionId, c.id.equals(activeId), c.session != null));
        }
        return out;
    }

    /** Activate a tab; returns its record, or null if the id is unknown. */
    public Conversation switchTo(String id) {
        if (!conversations.containsKey(id)) return null;
        activeId = id;
        return conversations.get(id);
    }

    /**
     * Remove a conversation. If it was active, activate the right neighbor (or
     * left if it was last; null if none remain) — the natural tab-close UX.
     */
    public CloseResult close(String id) {
        if (!conversations.containsKey(id)) {
            return new CloseResult(false, null, active());
        }
        int idx = order.indexOf(id);
        Conversation conv = conversations.remove(id);
        order.remove(idx);
        if (id.equals(activeId)) {
            if (idx < order.size()) {
                activeId = order.get(idx);          // former right neighbor
            } else if (idx - 1 >= 0) {
                activeId = order.get(idx - 1);      // was last → left neighbor
            } else {
                activeId = null;                    // none remain
            }
        }
        return new CloseResult(true, conv, active());
    }

    public Conversation setSessionId(String id, String sessionId) {
        Conversation c = get(id);
        if (c != null) c.sessionId = emptyToNull(sessionId);
        return c;
    }

    public Conversation setSession(String id, Object handle) {
        Conversation c = get(id);
        if (c != null) c.session = handle;
        return c;
    }

    /** Set a conversation's approval mode (takes effect on its next child spawn). */
    public Conversation setMode(String id, String mode) {
        Conversation c = get(id);
        if (c != null) c.mode = (mode == null || mode.trim().isEmpty()) ? "default" : mode;
        return c;
    }

    /** Set a conversation's extended-thinking level (next child spawn). */
    public Conversation setThinking(String id, String thinking) {
        Conversation c = get(id);
        if (c != null) c.thinking = (thinking == null || thinking.trim().isEmpty()) ? "off" : thinking;
        return c;
    }

    public Conversation setTitle(String id, String title) {
        Conversation c = get(id);
        if (c != null && title != null && !title.trim().isEmpty()) c.title = title;
        return c;
    }

    /** Fresh per-turn reducer state for a conversation (after reset/resume). */
    public Conversation resetTurnState(String id) {
        Conversation c = get(id);
        if (c != null) c.turnState = createTurnState.get();
        return c;
    }

    /** Every live session handle (for dispose — caller stops each). */
    public List<Object> allSessions() {
        List<Object> out = new ArrayList<Object>();
        for (String id : order) {
            Object s = conversations.get(id).session;
            if (s != null) out.add(s);
        }
        return out;
    }

    private static String emptyToNull(String s) {
        return (s == null || s.isEmpty()) ? null : s;
    }
}
